### ISSUE TITLE
CI builds of third-party branches time out and fail

### ISSUE THREAD

**Issue by geigerzaehler** [radicle-dev/radicle-upstream#1471]
All of our CI builds of third-party branches time out and fail. ([Example](https://buildkite.com/monadic/radicle-upstream/builds/6936#b81ce7d7-bc85-401a-8e56-7279f8e7f24c/157-265))

They hang at the following line:
~~~
test can_ask_and_clone_project ... fatal: '.' does not appear to be a git repository
~~~
Then they time out after 60 minutes.

This is probably due to us using the `kata` container runtime for third-party builds.

**Comment by geigerzaehler** [comment]
I’ve disabled third-party builds until we’ve fixed this to avoid clogging the queue. In the meantime, us maintainers need to push third-party contributions to our repo to build them.

**Comment by FintanH** [comment]
> In the meantime, us maintainers need to push third-party contributions to our repo to build them.

Can you elaborate on what you mean by this? As in don't fork?

> `test can_ask_and_clone_project ... fatal: '.' does not appear to be a git repository`

Any ideas as to why this might happen? I tried to look at the test before and nothing jumped out at me. It _looked_ really similar to every other test in there :/

**Comment by geigerzaehler** [comment]
I created a couple of [third-party builds](https://buildkite.com/monadic/radicle-upstream/builds?branch=geigerzaehler:fix-stall) with extended tracing. 

@kim suggested that the error “fatal: '.' does not appear to be a git repository” comes from git itself and indicates that the temporary directory that holds the repo has been deleted. Indeed we see from [the logs](https://buildkite.com/monadic/radicle-upstream/builds/6947#0dbdb06b-0024-480a-9ba8-75f793523869/73-303) that the `Peer` struct is dropped. Though I wasn’t able to determine the cause.

> > In the meantime, us maintainers need to push third-party contributions to our repo to build them.
> 
> Can you elaborate on what you mean by this? As in don't fork?

If somebody without push access to our repo creates a PR the CI build for that PR will fail. To check whether it passes one of us maintainers needs to pull the third-party branch and push it to our repo. (After verifying that it’s not malicious, of course.) We’ll use the result of that branches build then to determine if the original PR can be merged.

**Comment by geigerzaehler** [comment]
Here are some more details.

The builds reliably fail when using the Kata Container Runtime. Kata Containers is a runtime for running Docker containers. The runtime uses lightweight virtual machines for isolation. In our buildkite agents we use the kata runtime for third party builds (that is builds of pull request.) The reason for this is security. We want to put as many barriers between our build agents and untrusted code. Beyond the fact that the builds stall on kata runtimes I have not discovered anything.

I see the following options for next steps

1. Investigate what is causing this on the test side. There’s a good chance that something is wrong with the implementation or the tests and that the failures are just a symptom. In fact, there have been instances where this happens without the kata runtime ([#1378](https://github.com/radicle-dev/radicle-upstream/issues/1378)). The difficult part in approaching this would be to make the tests reproducable. One would need to run the tests in the kata runtime on their own machines. (Even then it’s not clear if it can be reproduced)
2. Continue the investigation into the kata runtime. Maybe we need to properly configure the runtime or the host machine? Maybe an update of the kata runtime helps? There are a lot of unknowns here which makes it hard to determine whether this is a good option.
3. Drop the kata container runtime and find a different solution to isolate third party builds from the build agent. Maybe rootless docker containers are indeed good enough? Maybe we can run trusted builds on dedicated machines?
4. Run third-party builds on hosted CI like Travis or Circle. This would require us to maintain two build setups but determining the actual effort requires more investigation. Maybe we can reuse a lot of the scripts? Hosted service require less effort to operate.

My recommendation would be to start with 1) and timebox the effort there. The next best option would be 4) unless there are big concerns about using another hosted service. 2) does not seem like a good option. 3) requires discussion and working out a solution.

**Comment by xla** [comment]
An additional option is to drop Buildkite and migrate to GH actions.

**Comment by geigerzaehler** [comment]
> An additional option is to drop Buildkite and migrate to GH actions

Yes, this would also be an option.

**Comment by kim** [comment]
I do think that the main reason for the suboptimal **cough** experience is not `kata-containers` per se (although they have some wobbles), but that we do not preserve caches for external builds. A cold cache build of our stack is just very slow.

Now,

* I agree that it is not Monadic's obligation to provide compute resources, and spend the time to do so securely
* Conversely, hosted CI solutions tend to make life more miserable for _everyone_ (including ourselves), because:
  * they are resource constrained due to virtualised hardware
  * even if you pay (which comes at a premium compared to on-premises hardware)
  * virtualisation adds significant build time overhead
  * they tend to not be customisable to the degree we would want (think: long-running nightlies, multi-process integration harnesses, ...)
  * security is dubious at best when producing release artifacts, or employing some sort of CD
  * and probably other things

In other words: you will need to grab my 2min build times from my cold dead hands!

That being said, it was always envisioned that a. builds from external contributors, and b. "smoke" tests on not-yet-commodity architectures could be farmed out, but it was not clear how to achieve this. It turns out that this is rather easy with [GH actions](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions) -- we simply need to untick "Build pull requests from third-party forked repositories" in BK, and add a conditional `if: github.repository_owner != "radicle-dev"` to GH.

So, **tl;dr** I think everyone wins by curating their build scripts into meaningful slices, and scheduling the "slow" set on BK, while scheduling only lint + unit test tasks on GH.

**Comment by kim** [comment]
https://github.com/radicle-dev/radicle-link/pull/530

It hurt only a little
